### DISCUSSION TITLE
How to check if a type is a TypeBuilder or contains a TypeBuilder generic parameter

### DISCUSSION THREAD

**Discussion by 2A5F** [dotnet/runtime#93330] [Q&A] [answered]
How to check if a type is a TypeBuilder or contains a TypeBuilder generic parameter

**Comment by KennethHoff** [comment]
```c#
var isTypeBuilder = x is TypeBuilder || x.GetType().GetGenericArguments().Any(a => a is TypeBuilder);
```

Does this work? If not, I don't think it's too far off - assuming I understood the question correctly.

**Reply by 2A5F** [reply]
![image](https://github.com/dotnet/runtime/assets/13982338/720b65de-1ba9-4809-84a3-848c57ac5af5)
Is there a way to check is TypeBuilderInstantiation directly, generics may be nested very deeply

**Reply by KennethHoff** [reply, marked as answer]
```c#
x is TypeBuilderInstantiation
```

the `is` pattern supports just mentioning the Type. I'm not too knowledgable about Reflection - hadn't even heard about `TypeBuilder` before this post.

You could recursively look through the `GetType().GetGenericArguments()` lists and then do something if any of them are whatever you need.

**Reply by 2A5F** [reply]
`TypeBuilderInstantiation` is internal, and most of its methods throw `NotSupportedException`, but fortunately `GetGenericArguments` does not

**Comment by 2A5F** [comment]
https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilderInstantiation.cs
